Allow array construction with mixed type shape tuples

ivirshup opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

This PR allows array constructor functions to accept shape tuples with heterogenous integer types.

I've done this by converting the contents of the shape tuple to integers of `types.intp` width, since that's what's expected downstream by `_empty_nd_impl` here:

https://github.com/numba/numba/blob/ad3303c25133d57127abf8be1cd72d6048f45813/numba/targets/arrayobj.py#L3339-L3341

Fixes #4575, and fixes #4529.

*Update* 

I've modified the original test to not include using a 64 bit integer in the shape tuple. This was due to errors in the 32 bit builds. They (of course) wouldn't allow extending a 64 bit integer to a 32 bit integer.

An alternative approach would be to see if we can safely downcast a value of greater than system width to system width. It might be good to check what numpy does if you've got an 32 bit system, and pass a shape tuple with 64 bit integers.

stuartarchibald commented:
Thanks for the PR. Perhaps also take a look at https://github.com/numba/numba/pull/3505, which does the same but also catches overflow based on type, it is however missing the run time check for the same.

ivirshup commented:
Thanks for the quick feedback @stuartarchibald! Totally forgot about your previous PR! Let me know if you'd like me to close this in favor of that one.

I've given a runtime error a shot, but don't have a 32 bit system handy to test on. I'll see what CI says I guess. What I had before this last commit would have just thrown a compilation error if a 64 bit int was passed.

stuartarchibald commented:
TBH, my previous PR stalled as I had to do other things, so someone else interested in moving it to a point where it can be merged is an excellent thing! Feel free to incorporate anything of value from it!!

The CI system will report ok for 32 bit systems. If you want to test locally you can also just create a "pretend" 32 bit system by using a 32bit Anaconda and setting some flags, pertinent bits of the build scripts:
https://github.com/numba/numba/blob/master/buildscripts/incremental/install_miniconda.sh#L8-L9
https://github.com/numba/numba/blob/master/buildscripts/incremental/setup_conda_environment.sh#L12-L13
https://github.com/numba/numba/blob/master/buildscripts/incremental/setup_conda_environment.sh#L58-L59

I also have a 32bit system in a VM as sometimes pretending isn't quite enough.

stuartarchibald commented:
Following a brief conversation with @sklam, it's probably best that to just check the allocation routines for null returns and raise if so, this also has the side effect of generally checking that memory was indeed allocated.

ivirshup commented:
If I pass too large of a shape to array creation, I currently just get a segfault, so I'm not sure if I'll get a return from allocators. I think I should do a 'safe' cast (no value change) to `intp` in `_parse_shape`, since downstream code requires `intp` shapes.

To better catch overflows, I could replicate [this check](https://github.com/numpy/numpy/blob/2a488fe76a0f732dc418d03b452caace161673da/numpy/core/src/multiarray/ctors.c#L1095-L1101) from numpy:

```C
        /*
         * Care needs to be taken to avoid integer overflow when
         * multiplying the dimensions together to get the total size of the
         * array.
         */
        if (npy_mul_with_overflow_intp(&nbytes, nbytes, dim)) {
            PyErr_SetString(PyExc_ValueError,
                "array is too big; `arr.size * arr.dtype.itemsize` "
                "is larger than the maximum possible size.");
            Py_DECREF(descr);
            return NULL;
        }
```

ivirshup commented:
I think this is working on all systems now. Thanks for the tips on getting a 32-bit conda env @stuartarchibald!

What else should I do here?

stuartarchibald commented:
Good job on getting this working cross platform :)

> If I pass too large of a shape to array creation, I currently just get a segfault, so I'm not sure if I'll get a return from allocators. I think I should do a 'safe' cast (no value change) to `intp` in `_parse_shape`, since downstream code requires `intp` shapes.
> 
> To better catch overflows, I could replicate [this check](https://github.com/numpy/numpy/blob/2a488fe76a0f732dc418d03b452caace161673da/numpy/core/src/multiarray/ctors.c#L1095-L1101) from numpy:
> 
> ```c
>         /*
>          * Care needs to be taken to avoid integer overflow when
>          * multiplying the dimensions together to get the total size of the
>          * array.
>          */
>         if (npy_mul_with_overflow_intp(&nbytes, nbytes, dim)) {
>             PyErr_SetString(PyExc_ValueError,
>                 "array is too big; `arr.size * arr.dtype.itemsize` "
>                 "is larger than the maximum possible size.");
>             Py_DECREF(descr);
>             return NULL;
>         }
> ```

I think implementing this check would be a good idea if you are interested in doing that? I think it'll be long term safer.

I'll have a dig around in the allocators and see if there's something that can be done there too (as this is a compiler, bypassing the NumPy routines for allocation is entirely possible!).

Thanks.

ivirshup commented:
> I think implementing this check would be a good idea if you are interested in doing that? I think it'll be long term safer.

I've implemented it, but it's definitely not comprehensive. My understanding is this check only covers "what could theoretically be addressable", but there are many lower values where numpy and malloc will refuse to allocate the array anyways:

```python
>>> np.zeros(1 << 44)
Python(40466,0x119e5f5c0) malloc: can't allocate region
*** mach_vm_map(size=140737488355328) failed (error code=3)
Python(40466,0x119e5f5c0) malloc: *** set a breakpoint in malloc_error_break to debug
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
MemoryError: Unable to allocate array with shape (17592186044416,) and data type float64
```

The exception raised by numpy changes depending on what the value is. Numba seems to segfault for this range of values. Right now I'm thinking the overflow checking I've added here is useful, but there should probably be more checks.

ivirshup commented:
@stuartarchibald, I've implemented the check for unaddressable arrays, was there anything else I should do here?

stuartarchibald commented:
> @stuartarchibald, I've implemented the check for unaddressable arrays, was there anything else I should do here?

This seems like a good approach. If you are happy I'll queue to for code review? Thanks.

stuartarchibald commented:
Bumped PR to get updated CI.

stuartarchibald commented:
@ivirshup this is failing flake8 here https://dev.azure.com/numba/numba/_build/results?buildId=3074&view=logs&jobId=687448c2-4012-5c9e-43b3-3520c9635a30&taskId=dba03d4a-57a7-5e59-a47a-79fe530f297a&lineStart=64&lineEnd=65&colStart=1&colEnd=1

stuartarchibald commented:
/AzurePipelines run

stuartarchibald commented:
@sklam RE getting this in for 0.47.0, I'm of the view that whilst there still needs to be some runtime checks done on the allocation returns, that this is a significant improvement, fixes a couple of known issues and should be considered for merge as-is (i.e. defer further work and merge this as it's useful in its own right).
